File: web/dashboard/src/services/auth.service.ts
import { apiClient } from '@/services/api-client'

export interface LoginRequest {
  email: string
  password: string
}

export interface RegisterRequest {
  email: string
  password: string
  firstName: string
  lastName: string
}

export interface AuthResponse {
  accessToken: string
  refreshToken: string
  expiresIn: number
  user: User
}

export interface User {
  id: string
  email: string
  firstName: string
  lastName: string
  role: 'ADMIN' | 'USER' | 'MERCHANT'
  enabled: boolean
  createdAt: string
}

class AuthApi {
  async login(request: LoginRequest): Promise<AuthResponse> {
    return apiClient.post<AuthResponse>('/auth/login', request)
  }

  async register(request: RegisterRequest): Promise<AuthResponse> {
    return apiClient.post<AuthResponse>('/auth/register', request)
  }

  async refreshToken(refreshToken: string): Promise<{ accessToken: string; refreshToken: string }> {
    return apiClient.post<{ accessToken: string; refreshToken: string }>('/auth/refresh', { refreshToken })
  }

  async logout(): Promise<void> {
    return apiClient.post<void>('/auth/logout')
  }

  async getCurrentUser(): Promise<User> {
    return apiClient.get<User>('/auth/me')
  }

  async updateProfile(data: Partial<User>): Promise<User> {
    return apiClient.patch<User>('/auth/profile', data)
  }

  async changePassword(currentPassword: string, newPassword: string): Promise<void> {
    return apiClient.post<void>('/auth/change-password', { currentPassword, newPassword })
  }
}

export const authApi = new AuthApi()

File: src/combined-service/src/main/java/dev/payment/combinedservice/auth/service/AuthService.java
package dev.payment.combinedservice.auth.service;

import dev.payment.combinedservice.auth.dto.AuthResponse;
import dev.payment.combinedservice.auth.dto.LoginRequest;
import dev.payment.combinedservice.auth.dto.RegisterRequest;
import dev.payment.combinedservice.auth.dto.UserResponse;
import dev.payment.combinedservice.auth.entity.User;
import dev.payment.combinedservice.auth.entity.UserPrincipal;
import dev.payment.combinedservice.auth.exception.AuthException;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    private final UserService userService;
    private final JwtService jwtService;
    private final PasswordEncoder passwordEncoder;
    private final TokenBlacklistService tokenBlacklistService;

    public AuthService(UserService userService, JwtService jwtService, 
                       PasswordEncoder passwordEncoder, TokenBlacklistService tokenBlacklistService) {
        this.userService = userService;
        this.jwtService = jwtService;
        this.passwordEncoder = passwordEncoder;
        this.tokenBlacklistService = tokenBlacklistService;
    }

    public AuthResponse register(RegisterRequest request) {
        User user = userService.createUser(
                request.email(),
                request.password(),
                request.firstName(),
                request.lastName(),
                request.role()
        );

        String accessToken = jwtService.generateAccessToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.of(
                accessToken,
                refreshToken,
                jwtService.getAccessTokenExpiration(),
                UserResponse.from(user)
        );
    }

    public AuthResponse login(LoginRequest request) {
        User user = userService.findByEmail(request.email())
                .orElseThrow(() -> new BadCredentialsException("Invalid email or password"));

        if (!passwordEncoder.matches(request.password(), user.getPasswordHash())) {
            throw new BadCredentialsException("Invalid email or password");
        }

        if (!user.isEnabled()) {
            throw new AuthException("Account is disabled", "ACCOUNT_DISABLED");
        }

        String accessToken = jwtService.generateAccessToken(user);
        String refreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.of(
                accessToken,
                refreshToken,
                jwtService.getAccessTokenExpiration(),
                UserResponse.from(user)
        );
    }

    public AuthResponse refreshToken(String refreshToken) {
        if (!jwtService.validateToken(refreshToken)) {
            throw new AuthException("Invalid refresh token", "INVALID_REFRESH_TOKEN");
        }

        String email = jwtService.extractEmail(refreshToken);
        User user = userService.findByEmail(email)
                .orElseThrow(() -> new AuthException("User not found", "USER_NOT_FOUND"));

        String newAccessToken = jwtService.generateAccessToken(user);
        String newRefreshToken = jwtService.generateRefreshToken(user);

        return AuthResponse.of(
                newAccessToken,
                newRefreshToken,
                jwtService.getAccessTokenExpiration(),
                UserResponse.from(user)
        );
    }

    public boolean validateToken(String token) {
        return jwtService.validateToken(token);
    }

    public User getUserByEmail(String email) {
        return userService.findByEmail(email)
                .orElseThrow(() -> new AuthException("User not found", "USER_NOT_FOUND"));
    }

    public void logout(String refreshToken) {
        if (refreshToken != null && !refreshToken.isBlank()) {
            try {
                if (jwtService.validateToken(refreshToken)) {
                    long expiration = jwtService.getRefreshTokenExpiration();
                    tokenBlacklistService.blacklistToken(refreshToken, expiration);
                }
            } catch (Exception e) {
                // Token invalid anyway, ignore
            }
        }
    }

    public UserResponse getCurrentUser(String email) {
        User user = userService.findByEmail(email)
                .orElseThrow(() -> new AuthException("User not found", "USER_NOT_FOUND"));
        return UserResponse.from(user);
    }
}
